### PR TITLE
PostgreSQL JDBC query timeouts

### PR DESCRIPTION
**Pull request by larshelge**
- Added timeout to JDBC template for 10 minutes.
- Added new JDBC template "slowQueryJdbcTemplate" for slow queries for 2 hours.
- Introduced QueryTimeoutException which can be used when wanting to catch timeouts, added handling to CrudControllerAdvice, so that API can return 409 conflict instead of 500 server error.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java [hunk 1]
```diff
@@ -1131,8 +1132,15 @@ private void fillGridWithAggregatedDataMap( DataQueryParams params, Grid grid, M
                 {
                     log.error( DebugUtils.getStackTrace( ex ) );
                     log.error( DebugUtils.getStackTrace( ex.getCause() ) );
-
-                    throw new RuntimeException( "Error during execution of aggregation query task", ex );
+                    
+                    if ( ex.getCause() != null && ex.getCause() instanceof RuntimeException )
+                    {
+                        throw (RuntimeException) ex.getCause(); // Throw the real exception instead of execution exception
```
Comment (huzvak) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java]: Maybe we should some log message as well? As we do in other classes? Just to be consistent. (If the exception is caught later and logged, then OK. But we should also log so it is easier to find out what is wrong just by looking into the log file)

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java [hunk 2]
```diff
@@ -109,7 +111,12 @@ public Grid getEvents( EventQueryParams params, Grid grid, int maxLimit )
         }
         catch ( BadSqlGrammarException ex )
         {
-            log.info( QUERY_ERR_MSG, ex );
+            log.info( AnalyticsUtils.ERR_MSG_TABLE_NOT_EXISTING, ex );
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: In this particular case the reason is that in the analytics engine design it is valid for some partitions not to exist - hence not using warn level as it is not an invalid state.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java (hunk 1 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java]: Yep here the exception is thrown / forwarded and hence logged. In this case, Spring wraps the exception in an ExecutionException as the method is invoked async. We are more interested in the underlying exception, hence throwing the "cause".

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 3]
```diff
@@ -243,7 +243,11 @@ else if ( maxLimit > 0 )
         }
         catch ( BadSqlGrammarException ex )
         {
-            log.info( QUERY_ERR_MSG, ex );
+            log.info( AnalyticsUtils.ERR_MSG_TABLE_NOT_EXISTING, ex );
+        }
+        catch ( DataAccessResourceFailureException ex )
+        {
+            log.info( AnalyticsUtils.ERR_MSG_QUERY_TIMEOUT, ex );
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]: Agree with you. Will fix.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: Agree we can raise log level to warn, will fix.

**Pull request merged**